src/go/rpk/pkg/cli/cmd/wasm/list_test.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package wasm
+
+import (
+	"bytes"
+	"encoding/json"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/Shopify/sarama"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/kafka"
+	kafkaMocks "github.com/vectorizedio/redpanda/src/go/rpk/pkg/kafka/mocks"
+)
+
+func newClientMock(offset int64) *kafkaMocks.MockClient {
+	return &kafkaMocks.MockClient{
+		MockGetOffset: func(string, int32, int64) (int64, error) {
+			return offset, nil
+		},
+	}
+}
+
+func parseLogrsOutput(output string) string {
+	plain := strings.TrimLeft(output, "level=info msg=\"")
+	plain = strings.Replace(plain, "\\", "", -1)
+	plain = strings.Replace(plain, "\n", "", -1)
+	plain = strings.TrimSuffix(plain, "\"")
+	return plain
+}
+
+func compareJsonResults(expected string, observed string) bool {
+	expectedParsed := make(map[string]interface{})
+	observedParsed := make(map[string]interface{})
+	if err := json.Unmarshal([]byte(expected), &expectedParsed); err != nil {
+		return false
+	}
+	if err := json.Unmarshal([]byte(observed), &observedParsed); err != nil {
+		return false
+	}
+	return reflect.DeepEqual(expectedParsed, observedParsed)
+}
+
+func TestNewListCommand(t *testing.T) {
+	tests := []struct {
+		name           string
+		args           []string
+		mockClient     kafkaMocks.MockClient
+		expectedOutput string
+		numberOfEvents int64
+		consumer       kafkaMocks.MockConsumer
+		pc             kafkaMocks.MockPartitionConsumer
+	}{
+		{
+			name:           "It should return the actual results",
+			args:           []string{"--raw"},
+			numberOfEvents: 1,
+			expectedOutput: "{\"0\": {\"status\":\"up\",\"node_id\":5,\"coprocessors\":{\"22220-name\":{\"input_topics\":[\"bar\"],\"description\":\"Hello World!\"}, \"77778-name\":{\"input_topics\":[\"baz\",\"foo\"],\"description\":\"Hello world again!\"}}}}",
+			pc: kafkaMocks.MockPartitionConsumer{
+				MockMessages: func() <-chan *sarama.ConsumerMessage {
+					messages := make(chan *sarama.ConsumerMessage)
+					go func() {
+						message := &sarama.ConsumerMessage{
+							Key:       []byte("0"),
+							Value:     []byte("{\"status\":\"up\",\"node_id\":5,\"coprocessors\":{\"22220-name\":{\"input_topics\":[\"bar\"],\"description\":\"Hello World!\"}, \"77778-name\":{\"input_topics\":[\"baz\",\"foo\"],\"description\":\"Hello world again!\"}}}"),
+							Topic:     kafka.CoprocessorStatusTopic,
+							Partition: 0,
+							Offset:    0,
+						}
+						messages <- message
+					}()
+					return messages
+				},
```

Comment:
You don't need to create a goroutine here. This could be simplified to:
```suggestion
				MockMessages: func() <-chan *sarama.ConsumerMessage {
					messages := make(chan *sarama.ConsumerMessage, 1)
					message := &sarama.ConsumerMessage{
						Key:       []byte("0"),
						Value:     []byte("{\"status\":\"up\",\"node_id\":5,\"coprocessors\":{\"22220-name\":{\"input_topics\":[\"bar\"],\"description\":\"Hello World!\"}, \"77778-name\":{\"input_topics\":[\"baz\",\"foo\"],\"description\":\"Hello world again!\"}}}"),
						Topic:     kafka.CoprocessorStatusTopic,
						Partition: 0,
						Offset:    0,
					}
					messages <- message
					return messages
				},
```

Reply:
You sure? I tried implementing this and the test stalls. I think what happens is that `list.go` is blocked on waiting on data from the topic, but this code is never called because it is blocked.

Reply:
But did you try *this* specific snippet? An important part is the buffered chan: `make(chan *sarama.ConsumerMessage, 1)`.

If you make a chan without a buffer, it will block when sending an element until another goroutine starts receiving. With a buffer (size 1, since you're only sending one message), it will send it without blocking and whenever another part of the code starts receiving, it will get the buffered message first.

Reply:
Ah I did not notice that, good point! Will include this change !